README.rst
flask-geckoboard
=================

Geckoboard_ is a hosted, real-time data dashboard that collects business
information from many online services and customers own data sources,
summarises it to distill the key messages and displays it in a way that's
easy to understand and interpret. It can be connected to virtually any source
of quantitative data.

This Flask_ extension provides view decorators to help create custom
widgets for geckoboard.

.. _Flask: http://flask.pocoo.org/
.. _Geckoboard: http://www.geckoboard.com/


CHANGELOG.rst
-Changelog
-=========
-
-Version 0.2.1
--------------
-* Make `comparative` field in Bullet Widget optional.
-
-Version 0.2.0
--------------
-* Bug fix in Bullet Widget.
-* Allow Bullet Widget to return more than one Bullet Widget.
-
-Version 0.1.0
--------------
-* First release, split off from django-geckoboard_.
-
-.. _django-analytical: http://pypi.python.org/pypi/django-geckoboard


tests/__init__.py
"""
Tests for flask-geckoboard.
"""

from test_decorators import *


tests/settings.py
"""
flask-geckoboard testing settings.
"""

GECKOBOARD_PASSWORD = 'pass123'


flask_geckoboard/__init__.py
"""
=================
flask-geckoboard
=================

Geckoboard_ is a hosted, real-time status board serving up indicators
from web analytics, CRM, support, infrastructure, project management,
sales, etc.  It can be connected to virtually any source of quantitative
data.  This Flask_ plugin provides view decorators to help create
custom widgets.

.. _Geckoboard: http://www.geckoboard.com/
.. _Flask: http://flask.pocoo.org/


Installation
============

To install flask-geckoboard, simply place the ``flask_geckoboard``
package somewhere on the Python path.


Limiting access
===============

If you want to protect the data you send to Geckoboard from access by
others, you can use an API key shared by Geckoboard and your widget
views.  Set ``GECKOBOARD_API_KEY`` in the Flask config.

If you do not set an API key, anyone will be able to view the data by
visiting the widget URL.


Encryption
==========

Geckoboard encryption allows encrypting data before it is sent to Geckoboard's
servers. After entering the password used to encrypt the data when the Geckoboard
is loaded, the data will be decrypted in the browser.

To use encryption, first set a ``GECKOBOARD_PASSWORD`` in the Flask config.

Next, enable encryption for each widget using the decorator arguments::

    from flask import Flask
    from flask_geckoboard import Geckoboard

    app = Flask(__name__)
    geckoboard = Geckoboard(app)

    @app.route('/user-count')
    @geckoboard.number(encrypted=True)
    def user_count(request):
        return User.objects.count()


Creating custom widgets
=======================

The available custom widgets are described in the Geckoboard support
section, under `Geckoboard API`_.  From the perspective of a Flask
project, a custom widget is just a view.  The flask-geckoboard
application provides view decorators that render the correct response
for the different widgets.

Let's say you want to add a widget to your dashboard that shows the
number of number of comments posted today.  First create a view, using a
flask-geckoboard decorator::

    from datetime import date, time, datetime

    from flask import Flask
    from flask_geckoboard import Geckoboard

    app = Flask(__name__)
    geckoboard = Geckoboard(app)

    @app.route('/comment-count')
    @geckoboard.number
    def comment_count():
        midnight = datetime.combine(date.today(), time.min)
        return Comment.objects.filter(submit_date__gte=midnight).count()


If your widget has optional settings, you can pass them in the decorator
definition::

    @app.route('/gecko/comment/count-absolute')
    @geckoboard.number(absolute='true')
    def comment_count(request):
        midnight = datetime.combine(date.today(), time.min)
        return Comment.objects.filter(submit_date__gte=midnight).count()


This is all the Flask code you need to display the comment count on
your dashboard. When you create a custom widget in Geckoboard, enter the
following information:

Encryption
    Enable if the field is encrypted (see instructions above).

URL data feed
    The view URL.  In the example above this would be something like
    ``http://HOSTNAME/geckoboard/comment_count/``.

API key
    The content of the ``GECKOBOARD_API_KEY`` setting, if you have set
    it.

Widget type
    *Custom*

Feed format
    Either *XML* or *JSON*.  If you don't specify a format the decorators will
    automatically detect and output the correct format or default to XML
    if this is not enabled (by default the format isn't appended by
    Geckoboard as a parameter any more)

Request type
    Either *GET* or *POST*.  The view decorators accept both.


The following decorators are available from the
``flask_geckoboard.decorators`` module:


``number``
-----------------

Render a *Number & Secondary Stat* widget.

The decorated view must return a tuple *(current, [previous],
[prefix])* where the *current* parameter is the current value, optional
*previous* parameter is the previous value of the measured quantity and
the optional parameter *prefix* is the prefix used in Geckoboard widget.
If there is only one parameter you do not need to return it in a tuple.
For example, to render a widget that shows the number of users and the
difference from last week::

    from datetime import datetime, timedelta

    @geckoboard.number
    def user_count(request):
        last_week = datetime.now() - timedelta(weeks=1)
        users = User.objects
        last_week_users = users.filter(date_joined__lt=last_week)
        return (users.count(), last_week_users.count())

    @geckoboard.number
    def users_count_with_prefix(request):
        last_week = datetime.now() - timedelta(weeks=1)
        users = User.objects
        last_week_users = users.filter(date_joined__lt=last_week)
        return (users.count(), last_week_users.count(), '$')


``rag``
--------------

Render a *RAG Column & Numbers* or *RAG Numbers* widget.

The decorated view must return a tuple with three tuples *(value,
[text])*.  The *value* parameters are the numbers shown in red, amber
and green (in that order).  The optional *text* parameters will be
displayed next to the respective values in the dashboard.

For example, to render a widget that shows the number of comments that
were approved or deleted by moderators in the last 24 hours::

    from datetime import datetime, timedelta

    @geckoboard.rag
    def comments(request):
        start_time = datetime.now() - timedelta(hours=24)
        comments = Comment.objects.filter(submit_date__gt=start_time)
        total_count = comments.count()
        approved_count = comments.filter(
                flags__flag=CommentFlag.MODERATOR_APPROVAL).count()
        deleted_count = Comment.objects.filter(
                flags__flag=CommentFlag.MODERATOR_DELETION).count()
        pending_count = total_count - approved_count - deleted_count
        return (
            (deleted_count, "Deleted comments"),
            (pending_count, "Pending comments"),
            (approved_count, "Approved comments"),
        )


``text``
---------------

Render a *Text* widget.

The decorated view must return a list of tuples *(message, [type])*.
The *message* parameters are strings that will be shown in the widget.
The *type* parameters are optional and tell Geckoboard how to annotate
the messages.  Use ``TEXT_INFO`` for informational messages,
``TEXT_WARN`` for for warnings and ``TEXT_NONE`` for plain text (the
default).  If there is only one plain message, you can just return it
without enclosing it in a list and tuple.

For example, to render a widget showing the latest Geckoboard twitter
updates, using Mike Verdone's `Twitter library`_::

    import twitter

    @geckoboard.text
    def twitter_status(request):
        twitter = twitter.Api()
        updates = twitter.GetUserTimeline('geckoboard')
        return [(u.text, TEXT_NONE) for u in updates]

.. _`Twitter library`: http://pypi.python.org/pypi/twitter


``pie_chart``
-------------

Render a *Pie chart* widget.

The decorated view must return an iterable over tuples *(value, label,
[color])*.  The optional *color* parameter is a string ``'RRGGBB[TT]'``
representing red, green, blue and optionally transparency.

For example, to render a widget showing the number of normal, staff and
superusers::

    @geckoboard.pie_chart
    def user_types(request):
        users = User.objects.filter(is_active=True)
        total_count = users.count()
        superuser_count = users.filter(is_superuser=True).count()
        staff_count = users.filter(is_staff=True,
                                   is_superuser=False).count()
        normal_count = total_count = superuser_count - staff_count
        return [
            (normal_count,    "Normal users", "ff8800"),
            (staff_count,     "Staff",        "00ff88"),
            (superuser_count, "Superusers",   "8800ff"),
        ]


``line_chart``
--------------

Render a *Line chart* widget.

The decorated view must return a tuple *(values, x_axis, y_axis,
[color])*.  The *values* parameter is a list of data points.  The
*x-axis* parameter is a label string or a list of strings, that will be
placed on the X-axis.  The *y-axis* parameter works similarly for the
Y-axis.  If there are more than one axis label, they are placed evenly
along the axis.  The optional *color* parameter is a string
``'RRGGBB[TT]'`` representing red, green, blue and optionally
transparency.

For example, to render a widget showing the number of comments per day
over the last four weeks (including today)::

    from datetime import date, timedelta

    @geckoboard.line_chart
    def comment_trend(request):
        since = date.today() - timedelta(days=29)
        days = dict((since + timedelta(days=d), 0)
                for d in range(0, 29))
        comments = Comment.objects.filter(submit_date__gte=since)
        for comment in comments:
            days[comment.submit_date.date()] += 1
        return (
            days.values(),
            [days[i] for i in range(0, 29, 7)],
            "Comments",
        )


``geck_o_meter``
----------------

Render a *Geck-O-Meter* widget.

The decorated view must return a tuple *(value, min, max)*.  The *value*
parameter represents the current value.  The *min* and *max* parameters
represent the minimum and maximum value respectively.  They are either a
value, or a tuple *(value, text)*.  If used, the *text* parameter will
be displayed next to the minimum or maximum value.

For example, to render a widget showing the number of users that have
logged in in the last 24 hours::

    from datetime import datetime, timedelta

    @geckoboard.geck_o_meter
    def login_count(request):
        since = datetime.now() - timedelta(hours=24)
        users = User.objects.filter(is_active=True)
        total_count = users.count()
        logged_in_count = users.filter(last_login__gt=since).count()
        return (logged_in_count, 0, total_count)


``funnel``
----------

Render a *Funnel* widget.

The decorated view must return a dictionary with at least an *items*
key.  To render a funnel showing the breakdown of authenticated users
vs. anonymous users::

    @geckoboard.funnel
    def user_breakdown(request):
        all_users = User.objects
        active_users =all_users.filter(is_active=True)
        staff_users = all_users.filter(is_staff=True)
        super_users = all_users.filter(is_superuser=True)
        return {
            "items": [
                (all_users.count(), 'All users'),
                (active_users.count(), 'Active users'),
                (staff_users.count(), 'Staff users'),
                (super_users.count(), 'Super users'),
            ],
            "type": "standard",   # default, 'reverse' changes direction
                                  # of the colors.
            "percentage": "show", # default, 'hide' hides the percentage
                                  # values.
            "sort": False,        # default, `True` orders the values
                                  # descending.
        }

``bullet``
----------

Render a *Bullet* widget.

The decorated view must return a dictionary with at least keys *label*,
*axis_points*, *current* and *comparative*. To render Geckoboard's own example
at
http://support.geckoboard.com/entries/274940-custom-chart-widget-type-definitions::

    @geckoboard.bullet
    def geckoboard_bullet_example(request):
        return = {
            'label': 'Revenue 2011 YTD',
            'axis_points': [0, 200, 400, 600, 800, 1000],
            'current': 500,
            'comparative': 600,
            'sublabel': 'U.S. $ in thousands',
            'red': [0, 400],
            'amber': [401, 700],
            'green': [701, 1000],
            'projected': [100, 900],
            'auto_scale': False,
        }

.. _`Geckoboard API`: http://geckoboard.zendesk.com/forums/207979-geckoboard-api
"""


class Geckoboard(object):
    app = None
    from .decorators import bar, bullet, funnel
    from .decorators import geck_o_meter
    from .decorators import leaderboard
    from .decorators import line_chart_legacy
    from .decorators import line_chart
    from .decorators import pie_chart
    from .decorators import text_widget as text
    from .decorators import rag_widget as rag
    from .decorators import number_widget as number

    def __init__(self, app=None):
        if app:
            self.init_app(app)

    def init_app(self, app):
        self.api_key = app.config.get('GECKOBOARD_API_KEY')
        self.password = app.config.get('GECKOBOARD_PASSWORD')
        self.app = app

__author__ = "Rob Eroh"
__email__ = "rob@eroh.me"
__version__ = "0.2.1"
__copyright__ = "Copyright (C) 2011-2014 Rob Eroh and others"
__license__ = "MIT License"


flask_geckoboard/decorators.py
""" Geckoboard decorators. """

import base64
from collections import OrderedDict
from functools import wraps
import json
from types import ListType, TupleType

try:
    from Crypto.Cipher import AES
    from Crypto import Random
    from hashlib import md5
    encryption_enabled = True
except ImportError:
    encryption_enabled = False

from flask import abort, request
from flask import current_app as app

TEXT_NONE = 0
TEXT_INFO = 2
TEXT_WARN = 1


class WidgetDecorator(object):
    """
    Geckoboard widget decorator.

    The decorated view must return a data structure suitable for
    serialization to XML or JSON for Geckoboard. See the Geckoboard
    API docs or the source of extending classes for details.

    If the ``GECKOBOARD_API_KEY`` setting is used, the request must
    contain the correct API key, or a 403 Forbidden response is
    returned.

    If the ``encrypted`` argument is set to True, then the data will be
    encrypted using ``GECKOBOARD_PASSWORD`` (JSON only).

    """
    def __new__(cls, *args, **kwargs):
        obj = object.__new__(cls)
        obj._encrypted = None
        if 'encrypted' in kwargs:
            if not encryption_enabled:
                raise GeckoboardException(
                    'Use of encryption requires the pycrypto package. ' +
                    'This package can be installed manually or by enabling ' +
                    'the encryption feature during installation.'
                )
            obj._encrypted = kwargs.pop('encrypted')
        obj._format = None
        if 'format' in kwargs:
            obj._format = kwargs.pop('format')
        obj.data = kwargs
        try:
            return obj(args[0])
        except IndexError:
            return obj

    def __call__(self, view_func):
        @wraps(view_func)
        def decorated_view(*args, **kwargs):
            if not _is_api_key_correct():
                abort(403)
            view_result = view_func(*args, **kwargs)
            data = self._convert_view_result(view_result)
            try:
                self.data.update(data)
            except ValueError:
                self.data = data
            content, content_type = _render(
                self.data, self._encrypted, self._format)
            return app.response_class(content, mimetype=content_type)
        return decorated_view

    def _convert_view_result(self, data):
        # Extending classes do view result mangling here.
        return data

widget = WidgetDecorator


class NumberWidgetDecorator(WidgetDecorator):
    """
    Geckoboard Number widget decorator.

    The decorated view must return a tuple `(current, [previous])`, where
    `current` is the current value and `previous` is the previous value
    of the measured quantity.

    """
    def _convert_view_result(self, result):
        if not isinstance(result, (tuple, list)):
            result = [result]
        result = list(result)
        for k, v in enumerate(result):
            result[k] = v if isinstance(v, dict) else {'value': v}
        return {'item': result}

number_widget = NumberWidgetDecorator


class RAGWidgetDecorator(WidgetDecorator):
    """
    Geckoboard Red-Amber-Green (RAG) widget decorator.

    The decorated view must return a tuple with three tuples `(value,
    [text])`. The `value` parameters are the numbers shown in red,
    amber and green (in that order).  The `text` parameters are optional
    and will be displayed next to the respective values in the
    dashboard.

    """
    def _convert_view_result(self, result):
        items = []
        for elem in result:
            if not isinstance(elem, (tuple, list)):
                elem = [elem]
            item = OrderedDict()
            if elem[0] is None:
                item['value'] = ''
            else:
                item['value'] = elem[0]
            if len(elem) > 1:
                item['text'] = elem[1]
            items.append(item)
        return {'item': items}

rag_widget = RAGWidgetDecorator


class TextWidgetDecorator(WidgetDecorator):
    """
    Geckoboard Text widget decorator.

    The decorated view must return a list of tuples `(message, [type])`.
    The `message` parameters are strings that will be shown in the
    widget. The `type` parameters are optional and tell Geckoboard how
    to annotate the messages. Use ``TEXT_INFO`` for informational
    messages, ``TEXT_WARN`` for for warnings and ``TEXT_NONE`` for plain
    text (the default).

    """
    def _convert_view_result(self, result):
        items = []
        if not isinstance(result, (tuple, list)):
            result = [result]
        for elem in result:
            if not isinstance(elem, (tuple, list)):
                elem = [elem]
            item = OrderedDict()
            item['text'] = elem[0]
            if len(elem) > 1 and elem[1] is not None:
                item['type'] = elem[1]
            else:
                item['type'] = TEXT_NONE
            items.append(item)
        return {'item': items}

text_widget = TextWidgetDecorator


class PieChartWidgetDecorator(WidgetDecorator):
    """
    Geckoboard Pie chart decorator.

    The decorated view must return a list of tuples `(value, label,
    color)`. The color parameter is a string 'RRGGBB[TT]' representing
    red, green, blue and optionally transparency.

    """
    def _convert_view_result(self, result):
        items = []
        for elem in result:
            if not isinstance(elem, (tuple, list)):
                elem = [elem]
            item = OrderedDict()
            item['value'] = elem[0]
            if len(elem) > 1:
                item['label'] = elem[1]
            if len(elem) > 2:
                item['colour'] = elem[2]
            items.append(item)
        return {'item': items}

pie_chart = PieChartWidgetDecorator


class LineChartLegacyWidgetDecorator(WidgetDecorator):
    """
    Geckoboard Line chart (legacy) decorator.

    The decorated view must return a tuple `(values, x_axis, y_axis,
    [color])`. The `values` parameter is a list of data points. The
    `x-axis` parameter is a label string or a list of strings, that will
    be placed on the X-axis. The `y-axis` parameter works similarly for
    the Y-axis. If there are more than one axis label, they are placed
    evenly along the axis. The optional `color` parameter is a string
    ``'RRGGBB[TT]'`` representing red, green, blue and optionally
    transparency.

    """
    def _convert_view_result(self, result):
        data = OrderedDict()
        data['item'] = list(result[0])
        data['settings'] = OrderedDict()

        if len(result) > 1:
            x_axis = result[1]
            if x_axis is None:
                x_axis = ''
            if not isinstance(x_axis, (tuple, list)):
                x_axis = [x_axis]
            data['settings']['axisx'] = x_axis

        if len(result) > 2:
            y_axis = result[2]
            if y_axis is None:
                y_axis = ''
            if not isinstance(y_axis, (tuple, list)):
                y_axis = [y_axis]
            data['settings']['axisy'] = y_axis

        if len(result) > 3:
            data['settings']['colour'] = result[3]

        return data

line_chart_legacy = LineChartLegacyWidgetDecorator


class LineChartWidgetDecorator(WidgetDecorator):
    """
    Geckoboard Line chart decorator.

    The decorated view must return a dict that contains at least a
    `series` entry which must be a list of dicts (one for each line) each
    containing a `data` entry with a list of the data to be plotted.

    Optional keys:

        x_axis: A dict with a `labels` entry for the x-axis labels.
        y_axis: A dict with a `labels` entry for the y-axis labels.

    The `series`, `x_axis`, and `y_axis` entries can have other optional keys.
    See https://developer.geckoboard.com/#line-chart for more information.

    """
    def _convert_view_result(self, result):
        data = OrderedDict()
        if not isinstance(result['series'], list) or 'series' not in result:
            raise RuntimeError('Key "series" (list) is required')
        for s in result.get('series'):
            if not isinstance(s, dict) or 'data' not in s:
                raise RuntimeError('Series must contain "data" entry')
        data['series'] = result.get('series')

        if 'x_axis' in result:
            data['x_axis'] = result.get('x_axis')

        if 'y_axis' in result:
            data['y_axis'] = result.get('x_axis')

        return data

line_chart = LineChartWidgetDecorator


class GeckOMeterWidgetDecorator(WidgetDecorator):
    """
    Geckoboard Geck-O-Meter decorator.

    The decorated view must return a tuple `(value, min, max)`. The
    `value` parameter represents the current value. The `min` and `max`
    parameters represent the minimum and maximum value respectively.
    They are either a value, or a tuple `(value, text)`. If used, the
    `text` parameter will be displayed next to the minimum or maximum
    value.

    """
    def _convert_view_result(self, result):
        value, min, max = result
        data = OrderedDict()
        data['item'] = value
        data['max'] = OrderedDict()
        data['min'] = OrderedDict()

        if not isinstance(max, (tuple, list)):
            max = [max]
        data['max']['value'] = max[0]
        if len(max) > 1:
            data['max']['text'] = max[1]

        if not isinstance(min, (tuple, list)):
            min = [min]
        data['min']['value'] = min[0]
        if len(min) > 1:
            data['min']['text'] = min[1]

        return data

geck_o_meter = GeckOMeterWidgetDecorator


class FunnelWidgetDecorator(WidgetDecorator):
    """
    Geckoboard Funnel decorator.

    The decorated view must return a dictionary with at least an `items`
    entry: `{'items': [(100, '100 %'), (50, '50 %')]}`.

    Optional keys are:

        type:       'standard' (default) or 'reverse'. Determines the
                    order of the colours.
        percentage: 'show' (default) or 'hide'. Determines whether or
                    not the percentage value is shown.
        sort:       `False` (default) or `True`. Sort the entries by
                    value or not.

    """
    def _convert_view_result(self, result):
        data = OrderedDict()
        items = result.get('items', [])

        # sort the items in order if so desired
        if result.get('sort'):
            items.sort(reverse=True)

        data['item'] = [dict(zip(('value', 'label'), item)) for item in items]
        data['type'] = result.get('type', 'standard')
        data['percentage'] = result.get('percentage', 'show')
        return data

funnel = FunnelWidgetDecorator


class BulletWidgetDecorator(WidgetDecorator):
    """
    The decorated method must return a dictionary containing these keys:

    Required keys:
    label:          Main label, eg. "Revenue 2011 YTD".
    axis_points:    Points on the axis, eg. [0, 200, 400, 600, 800, 1000].
    current:        Current value range, eg. 500 or [100, 500]. A singleton
                    500 is internally converted to [0, 500].
    comparative:    Comparative value, eg. 600.

    Optional keys:
    orientation:    One of 'horizontal' or 'vertical'. Defaults to horizontal.
    sublabel:       Appears below main label.
    range:          Ordered list of color ranges:
                    [{'color': 'red', 'start': 0, 'end': 1},
                     {'color': 'amber', 'start': 1, 'end': 5},
                     {'color': 'green', 'start': 5, 'end': 10}]
                    Defaults are calculated from axis_points.
    projected:      Projected value range, eg. 900 or [100, 900]. A singleton
                    900 is internally converted to [0, 900].

    auto_scale:     If true then values will be scaled down if they
                    do not fit into Geckoboard's UI, eg. a value of 1100
                    is represented as 1.1. If scaling takes place the sublabel
                    is suffixed with that information. Default is true.

    """
    def _convert_view_result(self, results):
        # Check required keys. We do not do type checking since this level of
        # competence is assumed.
        if not isinstance(results, list):
            results = [results]
        items = []
        for result in results:
            for key in ('label', 'axis_points', 'current'):
                if key not in result:
                    raise RuntimeError('Key {} is required'.format(key))

            # Handle singleton current and projected
            current = result['current']
            projected = result.get('projected', None)
            if not isinstance(current, (ListType, TupleType)):
                current = [0, current]
            if (projected is not None) and not isinstance(
                    projected, (ListType, TupleType)):
                projected = [0, projected]

            # If red, amber and green are not *all* supplied calculate defaults
            axis_points = result['axis_points']
            _range = result.get('range', [])
            if not _range:
                if axis_points:
                    max_point = max(axis_points)
                    min_point = min(axis_points)
                    third = (max_point - min_point) / 3
                    range.append({'color': 'red',
                                  'start': min_point,
                                  'end': min_point + third - 1})
                    range.append({'color': 'amber',
                                  'start': min_point + third,
                                  'end': max_point - third - 1})
                    range.append({'color': 'red',
                                  'start': max_point - third,
                                  'end': max_point})
                else:
                    _range = [{'color': 'red', 'start': 0, 'end': 0},
                              {'color': 'amber', 'start': 0, 'end': 0},
                              {'color': 'green', 'start': 0, 'end': 0}]

            # Scan axis points for largest value and scale to avoid overflow in
            # Geckoboard's UI.
            auto_scale = result.get('auto_scale', True)
            if auto_scale and axis_points:
                scale_label_map = {1000000000: 'billions', 1000000: 'millions',
                                   1000: 'thousands'}
                scale = 1
                value = max(axis_points)
                for n in (1000000000, 1000000, 1000):
                    if value >= n:
                        scale = n
                        break

                # Little fixedpoint helper.
                # todo: use a fixedpoint library
                def scaler(value, scale):
                    return float('%.2f' % (value*1.0 / scale))

                # Apply scale to all values
                if scale > 1:
                    axis_points = [scaler(v, scale) for v in axis_points]
                    current = (scaler(current[0], scale),
                               scaler(current[1], scale))
                    if projected is not None:
                        projected = (scaler(projected[0], scale),
                                     scaler(projected[1], scale))
                    # red = (scaler(red[0], scale), scaler(red[1], scale))
                    # amber = (scaler(amber[0], scale), scaler(amber[1], scale))
                    # green = (scaler(green[0], scale), scaler(green[1], scale))
                    if 'comparative' in result:
                        result['comparative'] = scaler(
                            result['comparative'], scale)

                    # Suffix sublabel
                    sublabel = result.get('sublabel', '')
                    if sublabel:
                        result['sublabel'] = '{} ({})'.format(
                            sublabel, scale_label_map[scale])
                    else:
                        result['sublabel'] = scale_label_map[scale].capitalize()

            # Assemble structure
            data = dict(
                label=result['label'],
                axis=dict(point=axis_points),
                range=_range,
                measure=dict(current=dict(start=current[0], end=current[1])))
            if 'comparative' in result:
                data['comparative'] = dict(point=result['comparative'])

            # Add optional items
            if 'sublabel' in result:
                data['sublabel'] = result['sublabel']
            if projected is not None:
                data['measure']['projected'] = dict(start=projected[0],
                                                    end=projected[1])

            items.append(data)
        return dict(item=items,
                    orientation=result.get('orientation', 'horizontal'),)

bullet = BulletWidgetDecorator


class LeaderboardWidgetDecorator(WidgetDecorator):
    """
    Leaderboard decorator.

    The decorated view must return a tuple that looks like the following:

        (labels, values, previous_ranks, 'ascending')

    The first three entries are lists with the corresponding labels, values, and
    previous rank in the leaderboard. If 'ascending' is provided as the last
    entry, the leaderboard is sorted in ascending order, else descending.

    The only required entries are labels and values, hence the minimal returned
    tuple must be like this:

        (labels, values)

    """
    def _convert_view_result(self, result):
        data = OrderedDict()
        sort_arg = 'descending'
        if len(result) > 3:
            sort_arg = result[-1]
            result = result[:-1]
        zipped = zip(*result)
        if len(result) > 2:
            items = [{'label': x[0], 'value': x[1],
                      'previous_rank': x[2]} for x in zipped]
        else:
            items = [{'label': x[0], 'value': x[1]} for x in zipped]
        # Sort the items by value
        if sort_arg == 'ascending':
            data['items'] = sorted(items, key=lambda k: k['value'])
        else:
            data['items'] = sorted(items, key=lambda k: k['value'],
                                   reverse=True)
        return data

leaderboard = LeaderboardWidgetDecorator


def _is_api_key_correct():
    """ Return whether the Geckoboard API key on the request is correct. """
    api_key = app.config.get('GECKOBOARD_API_KEY')
    if api_key is None:
        return True
    auth = request.authorization
    if auth:
        if auth.type == 'basic':
            return auth.username == api_key and auth.password == 'X'
    return False


def _derive_key_and_iv(password, salt, key_length, iv_length):
    d = d_i = ''
    while len(d) < key_length + iv_length:
        d_i = md5(d_i + password + salt).digest()
        d += d_i
    return d[:key_length], d[key_length:key_length+iv_length]


def _encrypt(data):
    """ Equivalent to OpenSSL using 256 bit AES in CBC mode. """
    BS = AES.block_size
    pad = lambda s: s + (BS - len(s) % BS) * chr(BS - len(s) % BS)
    password = app.config.get('GECKOBOARD_PASSWORD')
    salt = Random.new().read(BS - len('Salted__'))
    key, iv = _derive_key_and_iv(password, salt, 32, BS)
    cipher = AES.new(key, AES.MODE_CBC, iv)
    encrypted = 'Salted__' + salt + cipher.encrypt(pad(data))
    return base64.b64encode(encrypted)


def _render(data, encrypted, format=None):
    """
    Render the data to Geckoboard. If the `format` parameter is passed
    to the widget it defines the output format. Otherwise the output
    format is based on the `format` request parameter.

    A `format` paramater of ``json`` or ``2`` renders JSON output, any
    other value renders XML.

    """
    return _render_json(data, encrypted)


def _render_json(data, encrypted=False):
    data_json = json.dumps(data)
    if encrypted:
        data_json = _encrypt(data_json)
    return data_json, 'application/json'


class GeckoboardException(Exception):
    """
    Represents an error with the Geckoboard decorators.

    """
